src/Juvix/Core/IR/Typechecker/Log.hs

```diff
@@ -0,0 +1,190 @@
+module Juvix.Core.IR.Typechecker.Log where
+
+import Juvix.Library as Lib
+import qualified Juvix.Core.Usage as Usage
+import Juvix.Core.IR.Typechecker.Types as Typed
+import qualified Juvix.Core.IR.Types as IR
+import qualified Data.Text as T
+
+
+type HasLogTC primTy primVal m =
+  HasWriter "typecheckerLog" [Log primTy primVal] m
+
+tellLogs :: HasLogTC primTy primVal m => [Log primTy primVal] -> m ()
+tellLogs = tell @"typecheckerLog"
+
+tellLog :: HasLogTC primTy primVal m => Log primTy primVal -> m ()
+tellLog msg = tellLogs [msg]
+
+
+data Log primTy primVal
+  = TermIntro
+      (Context primTy primVal)
+      (IR.Term primTy primVal)
+      (Annotation primTy primVal)
+  | ElimIntro (Context primTy primVal) (IR.Elim primTy primVal)
+  | Typechecked (IR.Term primTy primVal) (Typed.Annotation primTy primVal)
+  | TcError (TypecheckError primTy primVal)
+
+  | CheckingStar
+  | CheckingSigmaZero
+  | SigmaIsZero
+  | CheckingLevels
+  | LevelOK Natural Natural
+
+  | CheckingPi
+  | CheckingPiAnnIsStar (IR.Value primTy primVal)
+  | PiAnnIsStar (IR.Value primTy primVal)
+  | CheckingPiArg
+  | CheckingPiRes
+
+  | CheckingPrimTy
+
+  | CheckingLam
+  | LamAnnIsPi
+  | LamBodyWith Usage.T Usage.T
+
+  | CheckingElim
+
+  | InferringFree
+  | FoundFree (Typed.Annotation primTy primVal)
+
+  | InferringPrim
+
+  | InferringApp
+  | AppFunIsPi
+      (IR.Elim primTy primVal)  -- ^ function
+      Usage.T                   -- ^ usage in annotation
+      (IR.Value primTy primVal) -- ^ function type
+      (IR.Term primTy primVal)  -- ^ argument
+  | AppInferredAs Usage.T (IR.Value primTy primVal)
+
+  | InferringAnn
+  | CheckingAnnIsType (IR.Term primTy primVal)
+  | CheckingAnnTerm (IR.Term primTy primVal) Usage.T (IR.Term primTy primVal)
+  deriving (Eq, Show)
+
+describe :: (Show primTy, Show primVal) => Log primTy primVal -> Text
+describe (TermIntro ctx tm (Annotation σ ty)) = mconcat
```

Comment:
Tbh I think the final form of this code will be using a pretty printing library to produce legible output instead of the derived `Show` form (same for `TypecheckError`). I just left the messages more or less as they were for now, to come back to later